CT-1751 SSD (#698)
keboola/connection-docs: storage/byodb/snowflake-secure-data-sharing/index.md

---
title: Snowflake Securat Data Sharing
permalink: /storage/byodb/snowflake-secure-data-sharing/
---

* TOC
{:toc}

This guide explains how to share a database and its objects with one or more accounts by creating a share. Currently, Keboola supports only granting privileges on objects directly to a share, specifically for tables, external tables, views, secure views, and secure materialized views. For further reference, consult the [Snowflake documentation](https://docs.snowflake.com/en/user-guide/data-sharing-gs#grant-privileges-directly-to-a-share).

The process involves two roles: the Producer and the Consumer.

Producer: The Snowflake account that owns and shares the data.

> Consumer: The Snowflake account that accesses the shared data.

Reviewer: ```suggestion
Consumer: The Snowflake account that accesses the shared data. This account has to be used in Keboola as [BYODB](https://help.keboola.com/storage/byodb/#main-header)
```

```sql
CREATE SHARE <SHARE_NAME>;
```

### 2. Add Objects to the Share by Granting Privileges
Grant privileges using `GRANT <privilege> ... TO SHARE`. You need to provide:
* USAGE privilege on the database to be shared. 
* USAGE privilege on each database schema containing the objects to be shared. 
* SELECT privilege for specific objects in the shared schema:
  * Tables
  * External tables
  * Secure views
  * Secure materialized views

Example:
```sql
GRANT USAGE ON DATABASE <DATABASE_NAME> TO SHARE <SHARE_NAME>;
GRANT USAGE ON SCHEMA <DATABASE_NAME>.<SCHEMA_NAME> TO SHARE <SHARE_NAME>;
GRANT SELECT ON ALL TABLES IN DATABASE <DATABASE_NAME> TO SHARE <SHARE_NAME>;
GRANT SELECT ON ALL TABLES IN SCHEMA <DATABASE_NAME>.<SCHEMA_NAME> TO SHARE <SHARE_NAME>;
GRANT SELECT ON TABLE <DATABASE_NAME>.<SCHEMA_NAME>.<TABLE_NAME> TO SHARE <SHARE_NAME>;
```

To review the object grants, use:
```sql
SHOW GRANTS TO SHARE <SHARE_NAME>;
```

### 3. Add Accounts to the Share
Add one or more accounts to the share using `ALTER SHARE`. To verify, use `SHOW SHARES`.
```sql
ALTER SHARE <SHARE_NAME ADD ACCOUNTS=<ACCOUNT_NAME>;
```
```sql
SHOW SHARES;
```

> The share is now ready for consumption by the specified accounts. For detailed guidance, refer to [Create and configure shares](https://docs.snowflake.com/en/user-guide/data-sharing-provider).

Reviewer: jen me napadlo, ze kdyz pouzivame `Share` jako pojem (podstatne jmeno), tak bychom ho mohli psat s velkym

> The share is now ready for consumption by the specified accounts. For detailed guidance, refer to [Create and configure shares](https://docs.snowflake.com/en/user-guide/data-sharing-provider).

Reviewer 2: Ja vim, Jirko, ze je to lakavy, ale kdyz bychom psali s velkym pismenem vsechno, co pouzivame jako pojem, nedaly by se nase texty skoro cist. Vypadaly by jako noty na buben nebo smlouvy od pravniku ;-). I Snowflake to pouziva s malym pismenem, ja bych u toho taky zustala.

### Role Requirements
To perform these tasks, use the `ACCOUNTADMIN` role or a role with the `IMPORT SHARE` global privilege. For more details, see [Enable non-ACCOUNTADMIN roles to perform data sharing tasks](https://docs.snowflake.com/en/user-guide/security-access-privileges-shares).

### Limitations of Imported Databases
Imported databases have the following restrictions:
* They are read-only. Consumers can query data but cannot modify it or create new objects.
* Unsupported actions include:
  * Cloning imported databases, schemas, or tables.
  * Using Time Travel for imported databases or their objects.
  * Editing comments for imported databases.
* They cannot be re-shared with other accounts.
* They cannot be replicated.

### Viewing Available Shares
Use the web interface or SQL to view available shares:
```sql
SHOW SHARES;
```

```sql
DESC SHARE <PRODUCER_ACCOUNT>.<SHARE_NAME>;
```

### Create a Database from a Share
To consume shared data, create a database from the share:
```sql
CREATE DATABASE <CONSUMER_DATABASE_NAME> FROM SHARE <PRODUCER_ACCOUNT>.<SHARE_NAME>;
```

### Granting privileges on an imported database

> Currently, Keboola supports only shares with **direct access to database objects**. Shares with roles are not supported.

Reviewer: `Shares with roles` - to by chtelo link do https://docs.snowflake.com/en/user-guide/data-sharing-policy-protected-data

Author: jo, to chtělo, link přidán

To allow users to access shared objects, grant the `IMPORTED PRIVILEGES` privilege on the imported database to one or more roles in your account. A role can grant `IMPORTED PRIVILEGES` only if it:
* Owns the imported database (i.e., has the `OWNERSHIP` privilege).
* Has been granted the `MANAGE GRANTS` global privilege.

Example:
```sql
GRANT IMPORTED PRIVILEGES ON DATABASE <CONSUMER_DATABASE_NAME> TO <KEBOOLA_PROJECT_ROLE>;
```

> 

Reviewer: ```suggestion
Now the `<CONSUMER_DATABASE_NAME>` database contains all the listed schemas provided by Producer.
```

## Final Steps

> You can now register the schema in the newly created database as an external bucket in Keboola, enabling seamless data integration.

Reviewer: ```suggestion
You can now register the schema in the newly created database as an External dataset in Keboola, enabling seamless data integration. Use `<CONSUMER_DATABASE_NAME>` as Database name and schema present in this database. When registering, don't forget to cehck the `Is Secure Data Share?` checkbox.
```

Reviewer: btw @kbl-tomas-hegedus meli bychom se teda rozhodnout, zda tomu budeme dale oficialne rikat SSD - Snowflake Shared Database nebo SDS - Secure Data Share. Nebo jeste kombo SSDS - Snowflake Secure Data Share. Ale at uz teda mame jasnej naming. a hodit ho pak do naseho jargonu

Reviewer 3: SDS - Secure Data Share

```sql
SHOW SHARES;
```

```sql
DESC SHARE <PRODUCER_ACCOUNT>.<SHARE_NAME>;
```

### Create a Database from a Share
To consume shared data, create a database from the share:
```sql
CREATE DATABASE <CONSUMER_DATABASE_NAME> FROM SHARE <PRODUCER_ACCOUNT>.<SHARE_NAME>;
```

### Granting Privileges on an Imported Database
Keboola currently supports only shares with [**direct access to database objects**](https://docs.snowflake.com/en/user-guide/data-sharing-policy-protected-data). Shares with roles are not supported.

To allow users to access shared objects, grant the `IMPORTED PRIVILEGES` privilege on the imported database to one or more roles in your account. A role can grant `IMPORTED PRIVILEGES` only if it:
* Owns the imported database (i.e., has the `OWNERSHIP` privilege).
* Has been granted the `MANAGE GRANTS` global privilege.

Example:
```sql
GRANT IMPORTED PRIVILEGES ON DATABASE <CONSUMER_DATABASE_NAME> TO <KEBOOLA_PROJECT_ROLE>;
```

## Final Step
You can now register the schema in the newly created database as an external dataset in Keboola, enabling seamless data integration. Use `<CONSUMER_DATABASE_NAME>` as the database name and 

> the schema present in this database. When registering, don't forget to check the `Secure Data Share` checkbox.

Reviewer 2: "Is Secure Data Share?" zni zvlastne. Navrhuju pouzit jenom "Secure Data Share", jestli souhlasis :-). Tooltip u toho vysvetluje, za jakych okolnosti to maji lidi zaskrtnout. Slo by to? Jestli jo, mohl bys to prosim upravit i v UI? 
Dekuju!

Author: Já to v UI neupravím. cc @kbl-tomas-hegedus

Reviewer 3: Souhlas, vypadá to líp. Zadám ticket na UI.

Author: Takže tomu ofiko říkáme odteď SDS - Secure Data Share?

viz https://github.com/keboola/connection-docs/pull/698#discussion_r1846477797
